src/Distribution/Client/Dynamic/Query.hs

```diff
@@ -147,15 +159,31 @@ withTempWorkingDir act = do
   setCurrentDirectory pwd
   res <$ removeDirectoryRecursive tmp
 
+-- | Declare useful functions for binary reading
+fmap concat $ mapM declareFunction [
+  mkName "Data.ByteString.Lazy.readFile"
+  , mkName "Data.ByteString.Lazy.Char8.dropWhile"
+  , mkName "Data.ByteString.Lazy.tail"
+  , mkName "Data.Binary.decode"
+  ]
+
 generateSource :: Selector LocalBuildInfo o -> String -> FilePath -> Version -> IO String
-generateSource (Selector s) modName setupConfig version =
+generateSource (Selector s) modName setupConfig version = 
   return $ flip generateModule modName $ do
-    getLBI <- addDecl (Ident "getLBI") $
-                   applyE fmap' (read' <>. unlines' <>. applyE drop' 1 <>. lines' :: ExpG (String -> LocalBuildInfo))
-               <>$ applyE readFile' (expr setupConfig)
+    getLBI <- if version < Version [1,22,0] []
+                then -- read via Read instance
+                  addDecl (Ident "getLBI") $
+                               applyE fmap' (read' <>. unlines' <>. applyE drop' 1 <>. lines' :: ExpG (String -> LocalBuildInfo))
+                           <>$ applyE Language.Haskell.Generate.readFile' (expr setupConfig)
+                else -- read via Binary Instance
+                  addDecl (Ident "getLBI") $
+                               applyE fmap' (decode' <>. Distribution.Client.Dynamic.Query.tail' <>. applyE Distribution.Client.Dynamic.Query.dropWhile' (applyE notequal' (expr '\n')) :: ExpG (Data.ByteString.Lazy.ByteString -> LocalBuildInfo) )
```

Comment:
Is it possible to use the non-qualified names here? (`tail'` instead of `Distribution.Client.Dynamic.Query`)